flipperdevices/flipperzero-firmware
# Allow calibrating CC1101 crystal frequency

# What's new

I seem to have been particularly unlucky with my device because the crystal ZQ3, specified as 10ppm, is off by 11ppm or so at room temperature. With this commit I can get the carrier frequency to be within measurement accuracy of my equipment (NESDR Nano2+, which has a TCXO with 0.5ppm accuracy).

I was tracking down a frequency error that caused replays of sub-GHz captures to be ignored by my DUT; this error is about 20% of the total frequency offset, with the other 80% being, uh, user error. By the time I figured that out, I implemented most of the functionality already, so I filed a PR. Feel free to close if this functionality is not wanted.

There is no GUI. I thought about adding another scene to Sub-GHz>Test at first but it seems pretty complicated, so for now it's just the CLI. It seems reasonable to expose this in the GUI in some way at least since if this value is set incorrectly for some reason it could send someone on a wild goose chase.

The only way to set the calibration (aside from direct filesystem manipulation) is through `subghz calibrate` CLI command.

# Verification 

- Run `subghz tx_carrier 433920000`. Measure the peak frequency.
- Run `subghz calibrate`. This returns an error since there's no calibration data yet.
- Run `subghz calibrate 100`.
- Run `subghz calibrate`. This should return 100.
- Run `subghz tx_carrier 433920000`. Measure the peak frequency. It should be ~43 kHz higher.
- Run `subghz calibrate -100`.
- Run `subghz tx_carrier 433920000`. Measure the peak frequency. It should be ~43 kHz lower.
- Run `subghz calibrate`. This should return -100.
- Run `subghz calibrate 0` to revert back to normal.

# Checklist (For Reviewer)
